src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/CompletionItemSource.cs

```diff
@@ -0,0 +1,253 @@
+                insertionText,
+                roslynItem.SortText,
+                roslynItem.FilterText,
+                attributeImages);
+
+            item.Properties.AddProperty(RoslynItem, roslynItem);
+            item.Properties.AddProperty(MatchPriority, roslynItem.Rules.MatchPriority);
+            item.Properties.AddProperty(SelectionBehavior, roslynItem.Rules.SelectionBehavior);
+            return item;
+        }
+
+        private ImmutableArray<EditorCompletion.CompletionFilter> GetFilters(RoslynCompletionItem item, Dictionary<string, EditorCompletion.CompletionFilter> filterCache)
+        {
+            var listBuilder = new ArrayBuilder<EditorCompletion.CompletionFilter>();
+            foreach (var filter in CompletionItemFilter.AllFilters)
+            {
+                if (filter.Matches(item))
+                {
+                    if (filterCache.TryGetValue(filter.DisplayText, out var applicableFilter))
+                    {
+                        listBuilder.Add(applicableFilter);
+                    }
+                    else
+                    {
+                        var imageId = Shared.Extensions.GlyphExtensions.GetGlyph(filter.Tags).GetImageId();
+                        var itemFilter = new EditorCompletion.CompletionFilter(
+                            filter.DisplayText,
+                            filter.AccessKey.ToString(),
+                            new ImageElement(new ImageId(imageId.Guid, imageId.Id), "Temporary Automation Name"));
+                        filterCache[filter.DisplayText] = itemFilter;
+                        listBuilder.Add(itemFilter);
+                    }
+                }
+            }
+
+            return listBuilder.ToImmutableAndFree();
+        }
+
+        public async Task<object> GetDescriptionAsync(EditorCompletion.CompletionItem item, CancellationToken cancellationToken)
+        {
+            // TODO: Should string.Empty returns here be null to avoid an empty popup? https://github.com/dotnet/roslyn/issues/27420
+
+            if (!item.Properties.TryGetProperty<RoslynCompletionItem>(RoslynItem, out var roslynItem) ||
+                !item.Properties.TryGetProperty<ITextBuffer>(TriggerBuffer, out var triggerBuffer))
+            {
+                return string.Empty;
+            }
+
+            if (!Workspace.TryGetWorkspace(triggerBuffer.AsTextContainer(), out var workspace))
+            {
+                return string.Empty;
+            }
+
+            var documentId = workspace.GetDocumentIdInCurrentContext(triggerBuffer.AsTextContainer());
+            if (documentId == null)
+            {
+                return string.Empty;
+            }
+
+            var document = workspace.CurrentSolution.GetDocument(documentId);
+            if (document == null)
+            {
+                return string.Empty;
+            }
+
+            var service = (CompletionServiceWithProviders)document.GetLanguageService<CompletionService>();
```

Comment:
Same as next function below? Can we reuse?

Reply:
The method just below is unused. Removing it. :-)

---
In reply to: [208429831](https://github.com/dotnet/roslyn/pull/29016#discussion_r208429831) [](ancestors = 208429831)